electrum/gui/wizard.py

```diff
@@ -0,0 +1,319 @@
+            # finished
+            self.finished(wizard_data)
+            return (None, wizard_data, {})
+
+        nexteval = nav['next']
+        # simple string based next view
+        if isinstance(nexteval, str):
+            new_view = WizardViewState(nexteval, wizard_data, {})
+        else:
+            # handler fn based next view
+            nv = nexteval(wizard_data)
+            self._logger.debug(repr(nv))
+
+            # append wizard_data and params if not returned
+            if isinstance(nv, str):
+                new_view = WizardViewState(nv, wizard_data, {})
+            elif len(nv) == 1:
+                new_view = WizardViewState(nv[0], wizard_data, {})
+            elif len(nv) == 2:
+                new_view = WizardViewState(nv[0], nv[1], {})
+            else:
+                new_view = nv
+
+        self._stack.append(copy.deepcopy(self._current))
+        self._current = new_view
+
+        self._logger.debug(f'resolve_next view is {self._current.view}')
+        self._logger.debug('stack:' + repr(self._stack))
+
+        return new_view
+
+    def resolve_prev(self):
+        prev_view = self._stack.pop()
+        self._logger.debug(f'resolve_prev view is {prev_view}')
+        self._logger.debug('stack:' + repr(self._stack))
+        self._current = prev_view
+        return prev_view
+
+    # check if this view is the final view
+    def is_last_view(self, view, wizard_data):
+        assert view
+        assert view in self.navmap
+
+        nav = self.navmap[view]
+
+        if not 'last' in nav:
+            return False
+
+        lastnav = nav['last']
+        # bool literal
+        if isinstance(lastnav, bool):
+            return lastnav
+        elif callable(lastnav):
+            # handler fn based
+            l = lastnav(view, wizard_data)
+            self._logger.debug(f'view "{view}" last: {l}')
+            return l
+        else:
+            raise Exception('last handler for view {view} is not callable nor a bool literal')
+
+    def finished(self, wizard_data):
+        self._logger.debug('finished.')
+
+    def reset(self):
+        self.stack = []
+        self._current = WizardViewState(None, {}, {})
+
+class NewWalletWizard(AbstractWizard):
+
+    _logger = get_logger(__name__)
+
+    def __init__(self, daemon):
+        self.navmap = {
+            'wallet_name': {
+                'next': 'wallet_type'
+            },
+            'wallet_type': {
+                'next': self.on_wallet_type
+            },
+            'keystore_type': {
+                'next': self.on_keystore_type
+            },
+            'create_seed': {
+                'next': 'confirm_seed'
+            },
+            'confirm_seed': {
+                'next': 'wallet_password',
+                'last': self.last_if_single_password
+            },
+            'have_seed': {
+                'next': self.on_have_seed,
+                'last': self.last_if_single_password_and_not_bip39
```

Comment:
I guess this item should not have a 'last' method

Reply:
have_seed is the last page, if using an electrum seed and unified password is enabled. 
If it crashes, it's a bug.

Reply:
```
W | gui.qml.qeapp | file:///opt/electrum/electrum/gui/qml/components/wizard/WCWalletName.qml
W | gui.qml.qeapp | wizard data is now :{}
W | gui.qml.qeapp | wizard data is now :{"wallet_name":"wallet_1"}
W | gui.qml.qeapp | next view: wallet_type
W | gui.qml.qeapp | file:///opt/electrum/electrum/gui/qml/components/wizard/WCWalletType.qml
W | gui.qml.qeapp | wizard data is now :{"wallet_name":"wallet_1","wallet_type":"standard","seed_type":"segwit"}
W | gui.qml.qeapp | next view: keystore_type
W | gui.qml.qeapp | file:///opt/electrum/electrum/gui/qml/components/wizard/WCKeystoreType.qml
W | gui.qml.qeapp | wizard data is now :{"seed_type":"segwit","wallet_name":"wallet_1","wallet_type":"standard","keystore_type":"haveseed"}
W | gui.qml.qeapp | next view: have_seed
W | gui.qml.qeapp | file:///opt/electrum/electrum/gui/qml/components/wizard/WCHaveSeed.qml
E | gui.qml.ElectrumGui | Traceback (most recent call last):
  File "/opt/electrum/electrum/gui/qml/qewizard.py", line 39, in isLast
    return self.is_last_view(self._current.view, wdata)
  File "/opt/electrum/electrum/gui/wizard.py", line 119, in is_last_view
    l = lastnav(view, wizard_data)
  File "/opt/electrum/electrum/gui/wizard.py", line 185, in last_if_single_password_and_not_bip39
    return self.last_if_single_password(view, wizard_data) and not wizard_data['seed_type'] == 'bip39'
KeyError: 'seed_type'
NoneType: None
W | gui.qml.qeapp | file:///opt/electrum/electrum/gui/qml/components/wizard/WCWalletName.qml:20: TypeError: Cannot call method 'suggestWalletName' of null
W | gui.qml.qeapp | Must construct a QGuiApplication first.


```

Reply:
note: I already had a wallet open in the above traceback

Reply:
Hmm.. I suspect you're on an older branch. Is your HEAD e32af1230eea9b3b18dc9a28180afc1f30e83375 or 5562cf2683c75ccd30d90c87dff8bf4d043dd210 ?

Reply:
no, I am not on an older branch.
If I checkout your commit directly, the crash does indeed not occur.
However, if I merge your branch from master, then it occurs.

Reply:
Have you tested this with `single_password = True`?
You will not get the above traceback if it is not set, because of lazy evaluation.

Reply:
Even with single_password = True my testcases all seemed to manage to avoid this issue :)
I was able to reproduce it now, and it's fixed in 2adf03477109f774b92e87529b06fd3b71861c8c

Reply:
hmm, any idea why you did not have the issue? why is this commit fixing it?

Reply:
I had `single_password` config option enabled, but the two test wallets I had in my test-unified folder were unencrypted. This lead to` self._use_single_password = self.daemon.update_password_for_directory(old_password=password, new_password=password)` in qedaemon.py to disable single password behavior

Reply:
I see. thanks